hapi-base/src/main/java/ca/uhn/hl7v2/util/XMLUtils.java

```diff
@@ -59,43 +65,50 @@ public synchronized static <T> T getDOMImpl() {
     }
 
     @SuppressWarnings("unchecked")
-    public static <T> T getDOMImplUncached() {
-        try {
-            DOMImplementationRegistry registry = DOMImplementationRegistry.newInstance();
-            return (T) registry.getDOMImplementation("LS 3.0");
-        } catch (Exception e) {
-            throw new RuntimeException(e);
-        }
-    }
-
     public static Document parse(String s) {
-        return parse(s, false);
+        return parseDocument(new InputSource(new StringReader(s)), false);
     }
 
     public static Document parse(String s, boolean validateIfSchema) {
-        DOMImplementationLS impl = getDOMImpl();
-        LSInput input = impl.createLSInput();
-        input.setStringData(s);
-        return parse(input, validateIfSchema);
+        return parseDocument(new InputSource(new StringReader(s)), validateIfSchema);
     }
 
     public static Document parse(InputStream s, boolean validateIfSchema) {
-        DOMImplementationLS impl = getDOMImpl();
-        LSInput input = impl.createLSInput();
-        input.setByteStream(s);
-        return parse(input, validateIfSchema);
+        return parseDocument(new InputSource(s), validateIfSchema);
     }
 
-    private static Document parse(LSInput input, boolean validateIfSchema) {
-        DOMImplementationLS impl = getDOMImpl();
-        LSParser parser = impl.createLSParser(DOMImplementationLS.MODE_SYNCHRONOUS, null);
-        DOMConfiguration config = parser.getDomConfig();
-        config.setParameter("element-content-whitespace", false);
-        config.setParameter("namespaces", true);
-        config.setParameter("validate-if-schema", validateIfSchema);
-        return parser.parse(input);
+    public static Document parseDocument(InputSource theInputSource, boolean theValidating) {

```

Comment:
Question: for my own understanding, what is the essence of the fix?

Reply:
@pepsiofficial The very-old-but-still-going Saxon API that we use to parse XML documents has a bunch of different mechanisms you can use to convert a string containing XML into a parsed DOM tree.

What we're doing here is replacing the `LSParser` API with the `DocumentBuilder` API, and in the process adding configuration that is specific to DocumentBuilder which disallows fetching external entities during parse.

This change should be otherwise functionally identical.. I just have no idea how to configure LSParser in an XXE safe way.

Reply:
Sweet, thanks for the explanation!